src/signature.c

```diff
@@ -174,103 +174,68 @@ static void p11prov_sig_freectx(void *ctx)
-const unsigned char der_rsa_sha1[] = { DER_SEQ_RSA_SHA1 };
-
-#define DER_ECSHA1_SEQ_LEN 0x09
-#define DER_ECSHA1_LEN 0x07
-/* 1.2.840.10045.4 */
+/* iso(1) member-body(2) us(840) ansi-x962(10045) signatures(4) */
 #define DER_ANSIX962_SIG 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x04
+#define DER_ANSIX962_SIG_LEN 0x06
 
-#define DER_ECSHA2_SEQ_LEN 0x10
-#define DER_ECSHA2_LEN 0x08
-/* 1.2.840.10045.4.3 */
+/* ... ansi-x962(10045) signatures(4) ecdsa-with-SHA2(3) */
 #define DER_ANSIX962_SHA2_SIG DER_ANSIX962_SIG, 0x03
+#define DER_ANSIX962_SHA2_SIG_LEN (DER_ANSIX962_SIG_LEN + 1)
 
-#define DER_SEQ_ECDSA_SHA224 \
-    DER_SEQUENCE, DER_ECSHA1_SEQ_LEN, DER_OBJECT, DER_ECSHA1_LEN, \
-        DER_ANSIX962_SHA2_SIG, 0x01
-const unsigned char der_ecdsa_sha224[] = { DER_SEQ_ECDSA_SHA224 };
-#define DER_SEQ_ECDSA_SHA256 \
-    DER_SEQUENCE, DER_ECSHA1_SEQ_LEN, DER_OBJECT, DER_ECSHA1_LEN, \
-        DER_ANSIX962_SHA2_SIG, 0x02
-const unsigned char der_ecdsa_sha256[] = { DER_SEQ_ECDSA_SHA256 };
-#define DER_SEQ_ECDSA_SHA384 \
-    DER_SEQUENCE, DER_ECSHA1_SEQ_LEN, DER_OBJECT, DER_ECSHA1_LEN, \
-        DER_ANSIX962_SHA2_SIG, 0x03
-const unsigned char der_ecdsa_sha384[] = { DER_SEQ_ECDSA_SHA384 };
-#define DER_SEQ_ECDSA_SHA512 \
-    DER_SEQUENCE, DER_ECSHA1_SEQ_LEN, DER_OBJECT, DER_ECSHA1_LEN, \
-        DER_ANSIX962_SHA2_SIG, 0x04
-const unsigned char der_ecdsa_sha512[] = { DER_SEQ_ECDSA_SHA512 };
-#define DER_SEQ_ECDSA_SHA1 \
-    DER_SEQUENCE, DER_ECSHA1_SEQ_LEN, DER_OBJECT, DER_ECSHA1_LEN, \
-        DER_ANSIX962_SIG, 0x01
-const unsigned char der_ecdsa_sha1[] = { DER_SEQ_ECDSA_SHA1 };
-
-#define DER_NISTID_SEQ_LEN 0x0D
-#define DER_NISTID_LEN 0x09
-/* 2.16.840.1.101.3.4.3 */
+/* joint-iso-itu-t(2) country(16) us(840) organization(1) gov(101) csor(3)
+ * nistAlgorithms(4) sigAlgs(3) */
 #define DER_NIST_SIGALGS 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x03
+#define DER_NIST_SIGALGS_LEN 0x08
+
+/* clang-format off */
+#define DEFINE_DER_SEQ_SHA(bits, rsa_algid, ecdsa_algid) \
+    static const unsigned char der_rsa_sha##bits[] = { \
+        DER_SEQUENCE, DER_RSADSI_PKCS1_LEN+5, \
+            DER_OBJECT, DER_RSADSI_PKCS1_LEN+1, DER_RSADSI_PKCS1, rsa_algid, \
+            DER_NULL, 0, \
+    }; \
+    static const unsigned char der_ecdsa_sha##bits[] = { \
+        DER_SEQUENCE, DER_ANSIX962_SHA2_SIG_LEN+3, \
+            DER_OBJECT, DER_ANSIX962_SHA2_SIG_LEN+1, DER_ANSIX962_SHA2_SIG, ecdsa_algid, \
+    };
+
+#define DEFINE_DER_SEQ_SHA3(bits, rsa_algid, ecdsa_algid) \
+    static const unsigned char der_rsa_sha3_##bits[] = { \
+        DER_SEQUENCE, DER_NIST_SIGALGS_LEN+5, \
+            DER_OBJECT, DER_NIST_SIGALGS_LEN+1, DER_NIST_SIGALGS, rsa_algid, \
+            DER_NULL, 0 \
+    }; \
+    static const unsigned char der_ecdsa_sha3_##bits[] = { \
+        DER_SEQUENCE, DER_NIST_SIGALGS_LEN+3, \
+            DER_OBJECT, DER_NIST_SIGALGS_LEN+1, DER_NIST_SIGALGS, ecdsa_algid \
+    };
+
+static const unsigned char der_rsa_sha1[] = {
```

Comment:
In theory we could declare this via the ugly:
DEFINE_DER_SEQ_SHA(1, 0x05, 0x01);

Or am I missing something?

Reply:
The ecdsa entry has different OID prefix, `DER_ANSIX962_SHA2_SIG` vs `DER_ANSIX962_SIG`. 

Though, this could be parametrized in the macros, by e.g.  adding `#define DEFINE_DER_ALGID(name, oid, oid_len)` and `#define DEFINE_DER_ALGID_WITH_NULL(name, oid, oid_len)` helpers to construct the individual DER arrays.

Reply:
Nah, I was just amused as it seemed possible, but it doesn't make sense to  waste time on this.
We should probably just remove SHA-1, we are blocking it in newest distributions anyways ... but that will be in the future.